### ISSUE TITLE
Bug in compute_confidences.h

### ISSUE THREAD

**Issue by Citronnier** [CGAL/cgal#5924]
Issue details:

In function 

`template <typename Kernel> std::vector<std::size_t> Candidate_confidences<Kernel>::supporting_points(Face_descriptor face, const Polygon_mesh& mesh, const Point_set& point_set)`

the circulator was not incremented when the current point is snapped, triggering an infinite loop. 
```
 do {
        Halfedge_descriptor hd = *cir;
        Vertex_descriptor vd = mesh.target(hd);
        const Point& p = coords[vd];
        const Point2& q = supporting_plane->to_2d(p);

        // Removes duplicated vertices
        // The last point in the polygon
        if (!plg.is_empty()) {
                const Point2& r = plg[plg.size() - 1];
                
                // an infinite loop will be triggered if this if block is executed, since cir will never be incremented
                if (CGAL::squared_distance(q, r) < CGAL::snap_squared_distance_threshold<FT>())
                        continue;
        }
        plg.push_back(q);

        ++cir;
} while (cir != done);
```

CGAL version: 5.3, x64-windows

**Comment by sloriot** [comment]
@LiangliangNan do you time to look into this issue?

**Comment by LiangliangNan** [comment]
@sloriot Yes. I can do it!
The proposed changes seems to solve the issue caused by the special case, but I will do some test.

@Citronnier Can you send me an example data (the smaller the better) so I can reproduce the bug?
You can attach it here or simply a DropBox link, or by any means.

**Comment by Citronnier** [comment]
@LiangliangNan Sure. This file is in PLY format and is generated from a segmentation result of the original input of my program, which is a surface mesh. I believe it's face No. 41415 among the generated candidates that triggered the bug.

https://cloud.tsinghua.edu.cn/f/b35e178a678f4debb7b7/?dl=1

**Comment by LiangliangNan** [comment]
@Citronnier  I couldn't reproduce the infinite loop you mentioned in this issue with the provided data. I tested it on a Mac with Clang and I used the extracted planes provided in your data. 

By reading the source code of the function `Candidate_confidences::supporting_points(...)`, I can confirm that a `++cir;` should be inserted before `continue`. With this change, all existing tests (with all existing test data) work as before (just because that the program does't reach the inserted new line of code). 

One thing related to this issue is the performance of the solver. With your data, the algorithm generated 357112 candidate faces, and it resulted in a linear integer program with 1059608 variables. This is a HUGE program and existing open source solvers cannot solve it within a reasonable time window (at least for SCIP, but I don't know if any commercial solver can). I don't know if this gives you the feeling that the program has an "infinite loop" (or maybe this data is really a special case and the missing `++cir;` indeed causes an infinite loop).

How to make your data work?
I visualized your extracted planes and found that their are too many tiny planar segments in your data. These tiny segments don't capture meaningful structure of your object (i.e., the building). But the pairwise intersection of them yields the large number of unnecessary candidate faces and thus a huge linear integer program that is difficult to solve. I suggest you tune the parameters of your plane extraction algorithm so that only dominant planes are kept for the reconstruction.

@sloriot I agree with the suggested change in #5925 and I suggest you accept the pull request (seems closed?).

**Comment by Citronnier** [comment]
@LiangliangNan Thank you for the reply. I understand the scale of the program made it infeasible, and it takes indefinitely long for the solver to solve it, yet it is not the "infinite loop" I reported in this issue. The reason for the reproduction failure could be attributed to the precision loss of exporting memory variables to text file and importing them back again (on my machine, it generated 1,059,584 candidates, and since the program is deterministic in nature, I believe this could be the reason)? I could have provided you with the original source code and .off file - only if you wish, of course.

As for the refining of the input data, I have already started toying around with methods to optimize it - and that's what I meant to do in the beginning. I am grateful for your time and advice!

**Comment by LiangliangNan** [comment]
@Citronnier Thanks for the clarification! Can you please reopen #5925 so it can be considered to be merged into the code base?

@sloriot The regarding issue is indeed a logic error and can be solved by #5925. Please proceed.

Thanks all!
